### PR TITLE
Create a Zodiac compatible Gnosis module with Neapolitan minion logic

### PR DESCRIPTION
**Pull request by ipatka**

### CODE REVIEW

File: contracts/SafeMinion.sol [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+
+    function onERC1155BatchReceived(address, address, uint256[] calldata, uint256[] calldata, bytes calldata) external pure override returns(bytes4)
+    {
+        return bytes4(keccak256("onERC1155BatchReceived(address,address,uint256[],uint256[],bytes)"));
+    }
+    
+    //  -- Moloch Withdraw Functions --
+    function doWithdraw(address token, uint256 amount) public memberOnly {
+        moloch.withdrawBalance(token, amount); // withdraw funds from parent moloch
+        IERC20(token).transfer(address(executor), amount);
```
Comment (ipatka) [inline contracts/SafeMinion.sol:241]: should SAFE be applicant?

File: contracts/SafeMinion.sol [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+        require(daoSignature.magicValue != 0, "erc1271::invalid signature");
+        require(
+            daoSignature.signatureHash ==
+                keccak256(abi.encodePacked(signature)),
+            "erc1271::invalid signature hash"
+        );
+        return daoSignature.magicValue;
+    }
+
+    function sign(
```
Comment (ipatka) [inline contracts/SafeMinion.sol:282]: who is the signing address

File: contracts/SafeMinion.sol [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+    function executeAction(uint256 proposalId, bytes memory transactions)
+        external
+        returns (bool)
+    {
+        Action memory action = actions[proposalId];
+        require(!action.executed, ERROR_EXECUTED);
+        // Mark executed before doing any external stuff
+        actions[proposalId].executed = true;
+
+        if (action.memberOrModule) {
```
Comment (auryn-macmillan) [inline contracts/SafeMinion.sol:392]: Maybe I'm reading this wrong, but this logic appears to be duplicated from `modifier memberOrModule`, is it necessary?

File: contracts/SafeMinion.sol [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+    event ChangeOwner(address owner);
+    event SetModule(address module);
+
+    modifier memberOnly() {
+        require(isMember(msg.sender), ERROR_MEMBER_ONLY);
+        _;
+    }
+
+    modifier executorOnly() {
+        require(msg.sender == address(executor), ERROR_MEMBER_ONLY);
```
Comment (auryn-macmillan) [inline contracts/SafeMinion.sol:232]: wrong error message?

File: contracts/SafeMinion.sol [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+        bool[6] memory flags = moloch.getProposalFlags(_proposalId);
+        require(flags[0], ERROR_NOT_SPONSORED);
+
+        // if any of these branches are true, let action proceed before proposal is processed
+        if (flags[2]) {
+            // if proposal has passed dao return true
+            return true;
+        }
+
+        if (module != address(0) && msg.sender == module) {
```
Comment (auryn-macmillan) [inline contracts/SafeMinion.sol:472]: This check also happens in `executeAction()` right before it calls `isPassed()`, so it's probably redundant here, no?

File: contracts/SafeMinion.sol (hunk 4 above)
Comment (ipatka) [reply contracts/SafeMinion.sol:232]: yes good catch

File: contracts/SafeMinion.sol [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+        // TODO: withdraw any funds from dao first? may need to verify this on the front end
+        moloch = IMOLOCH(_moloch);
+        molochDepositToken = moloch.depositToken();
+        // verify that moloch address has a deposit token and it is not zero
+        require(molochDepositToken != address(0), ERROR_ZERO_DEPOSIT_TOKEN);
+        emit ChangeOwner(_moloch);
+        return true;
+    }
+
+    function setModule(address _module) external executorOnly returns (bool) {
```
[the remark this answers is not in the document]
Comment (ipatka) [reply contracts/SafeMinion.sol:454]: vestige from the previous model, removing it

**Pull request merged**